crates/engine/src/io.rs

```diff
@@ -1,356 +1,342 @@
-use cap_std::fs::File as CapFile;
-use std::{
-    collections::HashSet,
-    fmt::Debug,
-    fs::{File, OpenOptions},
-    io::{LineWriter, Write},
-    path::PathBuf,
-    sync::{Arc, RwLock, RwLockReadGuard},
-};
+use std::io::{Read, Write};
+
+use spin_manifest::CoreComponent;
+use wasi_cap_std_sync::WasiCtxBuilder;
 use wasi_common::{
     pipe::{ReadPipe, WritePipe},
     WasiFile,
 };
-use wasmtime_wasi::sync::file::File as WasmtimeFile;
-
-/// Prepares a WASI pipe which writes to a memory buffer, optionally
-/// copying to the specified output stream.
-pub fn redirect_to_mem_buffer(
-    follow: Follow,
-) -> (WritePipe<WriteDestinations>, Arc<RwLock<WriteDestinations>>) {
-    let immediate = follow.writer();
 
-    let buffer: Vec<u8> = vec![];
-    let std_dests = WriteDestinations { buffer, immediate };
-    let lock = Arc::new(RwLock::new(std_dests));
-    let std_pipe = WritePipe::from_shared(lock.clone());
+use crate::Component;
 
-    (std_pipe, lock)
+/// Configuration for the WASI stdio (stdin, stdout, stderr) of a component.
+#[derive(Clone)]
+pub struct ComponentStdio {
+    stdin: StdInput,
+    stdout: StdOutput,
+    stderr: StdOutput,
 }
 
-/// Which components should have their logs followed on stdout/stderr.
-#[derive(Clone, Debug)]
-pub enum FollowComponents {
-    /// No components should have their logs followed.
-    None,
-    /// Only the specified components should have their logs followed.
-    Named(HashSet<String>),
-    /// All components should have their logs followed.
-    All,
-}
-
-impl FollowComponents {
-    /// Whether a given component should have its logs followed on stdout/stderr.
-    pub fn should_follow(&self, component_id: &str) -> bool {
-        match self {
-            Self::None => false,
-            Self::All => true,
-            Self::Named(ids) => ids.contains(component_id),
+impl Default for ComponentStdio {
+    fn default() -> Self {
+        Self {
+            stdin: StdInput::Null,
+            stdout: StdOutput::Null,
```

Comment:
Is this null as in /dev/null?  Suggest using `Discard` if so

Reply:
I stole from https://doc.rust-lang.org/std/process/struct.Stdio.html#method.null but `Discard` would be fine too.

Reply:
Oh, if it's an existing convention then happy to stick with it!